rfc/cart-api-v1.md

```diff
@@ -0,0 +1,179 @@
+    * the client will generate a version 4 UUID string as dubbed as session ID (and the backend will treat the session ID as user ID) and the base64-encoded string of that UUID will be a bearer token.
+
+2. For CRUD endpoints,
+  * All CRUD endpoints take JSON request body on `POST` and `PUT`/`PATCH` requests.
+  * `id` is not required as either the ID will be auto-generated by the endpoint on `POST` request, or the ID is given on the URL and immutable on `PUT`/`PATCH` request.
+  * Generally, the response on `GET` request will return a single JSON object of the corresponding model if the ID is specified or the list of JSON object otherwise. *There will be some exceptions.*
+  * For `PUT` and `PATCH` requests, if `id` in the URL path and `id` in the request body are different, the endpoint may respond with **HTTP 412** without doing anything (preferred), or **HTTP 200** without updating the ID of the corresponding entity.
+
+3. Generally, all endpoints may respond different HTTP status codes according to the given scenario, unless specified otherwise.
+
+| HTTP Status Code | Condition | Notes |
+| --- | --- | --- |
+| 200 | Everything is ok. | Default status code |
+| 401 | The bearer token is invalid or not present. | - |
+| 404 | Resource not found | Used on `GET` and `PUT` only |
+| 405 | Method not allowed | AWS API Gateway will handle this automatically. |
+| 409 | Data integrity error, e.g., supposing that item `i_1` is in cart `c_1`, the client mistakenly request to delete item `i_1` from cart `c_2`. | This is more commonly used in `PUT` and `DELETE` requests. |
+| 410 | Resource no longer existed | Used on `DELETE` only |
+| 412 | Invalid input / precondition failed | Requires input validation |
+
+#### CRUD APIs for `Cart`
+
+| Method | Path | Description |
+| --- | --- | --- |
+| `POST` | `/resources/carts/` | Create a new (non-default) cart |
+| `GET` | `/resources/carts/` | Retrieve a list of carts |
+| `GET` | `/resources/carts/{id}` | Retrieve a single cart (see more detail below) |
+| `PUT` | `/resources/carts/{id}` | Update the cart (see more detail below) |
+| `DELETE` | `/resources/carts/{id}` | Delete a single cart (see more detail below) |
+
+#### `GET /resources/carts/{id}`
+
+Retrieve a single cart.
+
+##### Response
+
+Instead of returning `Cart` information, the response will be also returning the list of associated `CartItem`, as shown below:
+
+```javascript
+{
+    id: str, // Cart ID
+    cart_name: str,
+    default: bool,
+    items: List[CartItem] // Each CartItem object will not have `cart_id`.
+}
+```
+
+##### Default Cart
+
+The default cart is a special case and will be given back to the client ONLY IF `id` is `default` (all lowercase).
```

Comment:
Just to clarify `default` is only used for this endpoint right?  So we can't do `PUT /resources/carts/default` to edit the default cart

Reply:
nvm missed the line a few lines down